src/osd/PrimaryLogPG.cc

```diff
@@ -5854,6 +5856,26 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
 	  break;
 	}
 
+	if (osd_op.op.flags & CEPH_OSD_OP_FLAG_WITH_DEDUP &&
+	    result == -ENOENT && obs.exists && mname == "chunk_get") {
+	    ClassHandler::ClassMethod *method = cls->get_method("chunk_set");
+	    if (!method) {
+	      dout(10) << "call method " << "refcount.chunk_set" << " does not exist" << dendl;
+	      result = -EOPNOTSUPP;
+	      break;
+	    }
+	    cls_chunk_refcount_get_op call;
+	    cls_chunk_refcount_set_op ref_set_op;
+	    decode(call, indata);
+	    ref_set_op.refs.insert(call.source);
+	    indata.clear();
+	    encode(ref_set_op, indata);
+	    result = method->exec((cls_method_context_t)&ctx, indata, outdata);
+	    if (result < 0) {
+	      break;
+	    }
+	}
+
```

Comment:
i don't understand why this is helping.  and regardless, i don't think we should be special casing something like this in the generic OP_CALL case!

Reply:
@liewegas 
Let assume that we flush the chunked object.
Operations look like below

* op[2] = "write op", "chunk_get"

Write op will make a new object, but the object will not be stored on filesystem yet when chunk_get is called. So, chunk_get will fail and it receives ENOENT. In this case, we need to use chunk_set instead of chunk_get because chunk_get read existing value first and then add a new value.

To handle this case, I think there are two solutions. First is making a new OP such as CEPH_OSD_OP_WRITE_WITH_REF. Second is adding a special case as above.

What is your opinion?